test/tools/yulopti.cpp

```diff
@@ -162,85 +162,97 @@ class YulOpti
+	{
+		if (!parse(_source))
+			return false;
+		disambiguate();
+		OptimiserSuite{m_context}.runSequence(_steps, *m_ast);
+		cout << AsmPrinter{m_dialect}(*m_ast) << endl;
+		return true;
+	}
+
+	void runInteractive(string _source, bool _disambiguated = false)
+	{
+		bool disambiguated = _disambiguated;
 		while (true)
 		{
-			cout << "----------------------" << endl;
-			cout << source << endl;
-			if (!parse(source))
+			if (!parse(_source))
 				return;
-			set<YulString> reservedIdentifiers;
-			if (!disambiguated)
-			{
-				*m_ast = std::get<yul::Block>(Disambiguator(m_dialect, *m_analysisInfo)(*m_ast));
-				m_analysisInfo.reset();
-				m_nameDispenser = make_shared<NameDispenser>(m_dialect, *m_ast, reservedIdentifiers);
-				disambiguated = true;
-			}
-			map<char, string> const& abbreviationMap = OptimiserSuite::stepAbbreviationToNameMap();
+			disambiguated = disambiguated || (disambiguate(), true);
 			map<char, string> const& extraOptions = {
 				// QUIT starts with a non-letter character on purpose to get it to show up on top of the list
 				{'#', ">>> QUIT <<<"},
 				{',', "VarNameCleaner"},
 				{';', "StackCompressor"}
 			};
 
-			printUsageBanner(abbreviationMap, extraOptions, 4);
+			printUsageBanner(extraOptions, 4);
 			cout << "? ";
 			cout.flush();
-			// TODO: handle EOF properly.
 			char option = static_cast<char>(readStandardInputChar());
 			cout << ' ' << option << endl;
 
-			OptimiserStepContext context{
-				m_dialect,
-				*m_nameDispenser,
-				reservedIdentifiers,
-				solidity::frontend::OptimiserSettings::standard().expectedExecutionsPerDeployment
-			};
-
-			auto abbreviationAndName = abbreviationMap.find(option);
-			if (abbreviationAndName != abbreviationMap.end())
+			try
 			{
-				OptimiserStep const& step = *OptimiserSuite::allSteps().at(abbreviationAndName->second);
-				step.run(context, *m_ast);
+				switch (option)
+				{
+					case 4:
+					case '#':
+						return;
+					case ',':
+						VarNameCleaner::run(m_context, *m_ast);
+						// VarNameCleaner destroys the unique names guarantee of the disambiguator.
+						disambiguated = false;
+						break;
+					case ';':
+					{
+						Object obj;
+						obj.code = m_ast;
+						StackCompressor::run(m_dialect, obj, true, 16);
+						break;
+					}
+					default:
+						OptimiserSuite{m_context}.runSequence(
+							std::string_view(&option, 1),
+							*m_ast
+						);
+				}
+				_source = AsmPrinter{m_dialect}(*m_ast);
 			}
-			else switch (option)
-			{
-			case '#':
-				return;
-			case ',':
-				VarNameCleaner::run(context, *m_ast);
-				// VarNameCleaner destroys the unique names guarantee of the disambiguator.
-				disambiguated = false;
-				break;
-			case ';':
+			catch (exception const& _exception)
 			{
-				Object obj;
-				obj.code = m_ast;
-				StackCompressor::run(m_dialect, obj, true, 16);
-				break;
+				cerr << "Exception during optimiser step: " << _exception.what() << endl;
 			}
```

Comment:
I think we want more diagnostic info for assertions and random exceptions:
```suggestion
			catch (OptimizerException const& _exception)
			{
				cerr << "Exception during optimiser step: " << _exception.what() << endl;
			}
			catch (exception const& _exception)
			{
				cerr << "Unexpected exception during optimiser step:" << endl;
				cerr << boost::diagnostic_information(_exception) << endl;
			}
			catch (...)
			{
				cerr << "Unexpected exception during optimiser step:" << endl;
				cerr << boost::current_exception_diagnostic_information() << endl;
			}
```

Actually, maybe we can even skip the `exception` case (the `(...)` block should cover it) but I haven't checked that so to be sure I'd add both.

Reply:
I can go for the ``(...)`` case only - but the main reason for doing this is to catch assertions due to unmet requirements, which are "expected", but not ``OptimizerExceptions``, but failing asserts. But yeah, I can just use ``boost::current_exception_diagnostic_information()`` in an catch-all for it.

Reply:
E.g. things like ``Need to run the function grouper before the stack compressor.`` (a ``YulAssertion``) - stuff like that happens easily and it seems nicer to me to just report it and give you the chance to run the proper step up front rather than quitting or crashing.

Reply:
I'm not suggesting quitting here. Just printing more info than a simple `what()` will give you. Especially for asserts these often just have a generic message (and until #12019 they didn't even have that) so full info might be more helpful in a general case.

For the few cases where there is enough info in the message, maybe we should just replace them with `OptimizerException` (or a subclass)? That could also be a `good first issue`.